### ISSUE TITLE
Keys not correctly saved when using Localstack and client side compression

### ISSUE THREAD

**Issue by andrewlock** [hotchkj/AspNetCore.DataProtection.Aws#28]
Framework: ASP.NET Core 2.0 app
Package version: `2.0.0-beta02`
AWS back end: [Localstack](https://github.com/localstack/localstack) docker image (localstack/localstack:latest)

I realise this is probably a bit of an esoteric setup, but just thought I would flag it in case others are having a similar issue.

Localstack provides an AWS emulation stack, including S3, so I was trying to set it up locally to use as the backing store for the DataProtection KeyRing for an ASP.NET Core 2.0 app I'm porting. 

Long story short, the data protection repository was able to access the Localstack S3 bucket, and attempted to write the key file to the bucket. However, when it tries to read it again, I was getting an error from the `AwsS3Client` library along the lines of `Observed hash did not match expected hash` (I've misplaced the exact error now unfortunately.)

After a lot of trial and error, I established that the DataProtection repo was creating truncated files on S3. E.g. this is the entirety of the file:

```xml
<?xml version="1.0" encoding="utf-8"?>
<key id="94ff8aac-8c7f-4e22-95ff-bc2e439882a4" version="1">
  <creationDate>2017-09-18T12:50:12.7618834Z</creationDate>
  <activationDate>2017-09-18T12:50:12.5115955Z</activationDate>
  <expirationDate>2017-12-17T12:50:12.5115955Z</expirationDate>
  <descriptor deserializerType="Microsoft.AspNetCore.DataProtection.AuthenticatedEncryption.ConfigurationModel.AuthenticatedEncryptorDescriptorDeserializer, Microsoft.AspNetCore.DataProtection, Version=2.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60">
    <descriptor>
      <encryption algo
```

As you can see, it just cuts off, so the MD5 hash [written to the start of the file](https://github.com/hotchkj/AspNetCore.DataProtection.Aws/blob/master/src/AspNetCore.DataProtection.Aws.S3/S3XmlRepository.cs#L274) did not match the contents.

Interestingly, I solved this by disabling `ClientSideCompression`, so I wonder if there's some issue with flushing the stream when converting to GZIP. 

As I say, there's a lot of moving parts here, so the issue might be elsewhere (i.e. Localstack) but thought it would be worth mentioning. I can work around it for now, but will report back if I find anything else.

P.S. If you play with Localstack yourself, make sure you set `AmazonS3Config.ForcePathStyle=true`

**Comment by hotchkj** [comment]
Thanks for reporting this issue.

To be 100% sure I understand - the write succeeded, but the subsequent read failed? That is particularly strange, since the aim of [that code](https://github.com/hotchkj/AspNetCore.DataProtection.Aws/blob/master/src/AspNetCore.DataProtection.Aws.S3/S3XmlRepository.cs#L274) is that S3 should check the MD5 of the PUT data _before_ writing the final object - in other words if you make `pr.Md5Digest` be garbage, the PUT itself should fail, with `The Content-MD5 you specified did not match what we received.`

The error you describe sounds like it's from [the AWS SDK read code](https://github.com/aws/aws-sdk-net/blob/ae822fc19be5d95e38e777672f51c9358be99a6d/sdk/src/Services/S3/Custom/Internal/AmazonS3ResponseHandler.cs#L213) which is checking the downloaded file vs the ETag. The ETag should be correct & representative of the content.

I won't immediately have time to fire up Localstack to check this myself and it does appear to work with vanilla S3, but I do notice that this is not the first such issue reported in Localstack (https://github.com/localstack/localstack/issues/98 - listed as closed but looks suspiciously similar).

I wonder whether Localstack is interpreting the streamed data in some way, perhaps decompressing it before it reaches the AWS SDK...

**Comment by andrewlock** [comment]
Thanks, so yes, it looks like the write succeeds (no errors written to output), and it's only when reading the key again that the hash fails. I've just reproduced here to get the stack:

<details>
<summary>This is the output from the Console logger, no mention of error on read (click to view)</summary>
<pre><code>
2017-09-19 09:15:54 [Information] [] Creating key {b91c4305-a1f4-4c42-a646-da7f357c0da5} with creation date 2017-09-19 08:15:54Z, activation date 2017-09-19 08:15:54Z, and expiration date 2017-12-18 08:15:54Z.
2017-09-19 09:15:54 [Warning] [] No XML encryptor configured. Key {b91c4305-a1f4-4c42-a646-da7f357c0da5} may be persisted to storage in unencrypted form.
2017-09-19 09:15:54 [Error] [] An error occurred while reading the key ring.
System.AggregateException: One or more errors occurred. (Expected hash not equal to calculated hash) ---> Amazon.Runtime.AmazonClientException: Expected hash not equal to calculated hash
   at Amazon.Runtime.Internal.Util.HashStream.CalculateHash() in E:\JenkinsWorkspaces\v3-trebuchet-release\AWSDotNetPublic\sdk\src\Core\Amazon.Runtime\Internal\Util\HashStream.cs:line 267
   at Amazon.Runtime.Internal.Util.HashStream.Dispose(Boolean disposing) in E:\JenkinsWorkspaces\v3-trebuchet-release\AWSDotNetPublic\sdk\src\Core\Amazon.Runtime\Internal\Util\HashStream.cs:line 172
   at System.IO.Stream.Close()
   at System.IO.Compression.DeflateStream.Dispose(Boolean disposing)
   at System.IO.Stream.Close()
   at System.IO.Compression.GZipStream.Dispose(Boolean disposing)
   at System.IO.Stream.Close()
   at AspNetCore.DataProtection.Aws.S3.S3XmlRepository.<GetElementFromKey>d__12.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at AspNetCore.DataProtection.Aws.S3.S3XmlRepository.<GetAllElementsAsync>d__11.MoveNext()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at AspNetCore.DataProtection.Aws.S3.S3XmlRepository.GetAllElements()
   at Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager.GetAllKeys()
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.CreateCacheableKeyRingCore(DateTimeOffset now, IKey keyJustAdded)
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.CreateCacheableKeyRingCore(DateTimeOffset now, IKey keyJustAdded)
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.Microsoft.AspNetCore.DataProtection.KeyManagement.Internal.ICacheableKeyRingProvider.GetCacheableKeyRing(DateTimeOffset now)
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.GetCurrentKeyRingCore(DateTime utcNow)
---> (Inner Exception #0) Amazon.Runtime.AmazonClientException: Expected hash not equal to calculated hash
   at Amazon.Runtime.Internal.Util.HashStream.CalculateHash() in E:\JenkinsWorkspaces\v3-trebuchet-release\AWSDotNetPublic\sdk\src\Core\Amazon.Runtime\Internal\Util\HashStream.cs:line 267
   at Amazon.Runtime.Internal.Util.HashStream.Dispose(Boolean disposing) in E:\JenkinsWorkspaces\v3-trebuchet-release\AWSDotNetPublic\sdk\src\Core\Amazon.Runtime\Internal\Util\HashStream.cs:line 172
   at System.IO.Stream.Close()
   at System.IO.Compression.DeflateStream.Dispose(Boolean disposing)
   at System.IO.Stream.Close()
   at System.IO.Compression.GZipStream.Dispose(Boolean disposing)
   at System.IO.Stream.Close()
   at AspNetCore.DataProtection.Aws.S3.S3XmlRepository.<GetElementFromKey>d__12.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at AspNetCore.DataProtection.Aws.S3.S3XmlRepository.<GetAllElementsAsync>d__11.MoveNext()<---
2017-09-19 09:15:54 [Information] [] Key ring failed to load during application startup.
System.AggregateException: One or more errors occurred. (Expected hash not equal to calculated hash) ---> Amazon.Runtime.AmazonClientException: Expected hash not equal to calculated hash
   at Amazon.Runtime.Internal.Util.HashStream.CalculateHash() in E:\JenkinsWorkspaces\v3-trebuchet-release\AWSDotNetPublic\sdk\src\Core\Amazon.Runtime\Internal\Util\HashStream.cs:line 267
   at Amazon.Runtime.Internal.Util.HashStream.Dispose(Boolean disposing) in E:\JenkinsWorkspaces\v3-trebuchet-release\AWSDotNetPublic\sdk\src\Core\Amazon.Runtime\Internal\Util\HashStream.cs:line 172
   at System.IO.Stream.Close()
   at System.IO.Compression.DeflateStream.Dispose(Boolean disposing)
   at System.IO.Stream.Close()
   at System.IO.Compression.GZipStream.Dispose(Boolean disposing)
   at System.IO.Stream.Close()
   at AspNetCore.DataProtection.Aws.S3.S3XmlRepository.<GetElementFromKey>d__12.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at AspNetCore.DataProtection.Aws.S3.S3XmlRepository.<GetAllElementsAsync>d__11.MoveNext()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at AspNetCore.DataProtection.Aws.S3.S3XmlRepository.GetAllElements()
   at Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager.GetAllKeys()
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.CreateCacheableKeyRingCore(DateTimeOffset now, IKey keyJustAdded)
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.CreateCacheableKeyRingCore(DateTimeOffset now, IKey keyJustAdded)
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.Microsoft.AspNetCore.DataProtection.KeyManagement.Internal.ICacheableKeyRingProvider.GetCacheableKeyRing(DateTimeOffset now)
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.GetCurrentKeyRingCore(DateTime utcNow)
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.GetCurrentKeyRing()
   at Microsoft.AspNetCore.DataProtection.Internal.DataProtectionStartupFilter.Configure(Action`1 next)
---> (Inner Exception #0) Amazon.Runtime.AmazonClientException: Expected hash not equal to calculated hash
   at Amazon.Runtime.Internal.Util.HashStream.CalculateHash() in E:\JenkinsWorkspaces\v3-trebuchet-release\AWSDotNetPublic\sdk\src\Core\Amazon.Runtime\Internal\Util\HashStream.cs:line 267
   at Amazon.Runtime.Internal.Util.HashStream.Dispose(Boolean disposing) in E:\JenkinsWorkspaces\v3-trebuchet-release\AWSDotNetPublic\sdk\src\Core\Amazon.Runtime\Internal\Util\HashStream.cs:line 172
   at System.IO.Stream.Close()
   at System.IO.Compression.DeflateStream.Dispose(Boolean disposing)
   at System.IO.Stream.Close()
   at System.IO.Compression.GZipStream.Dispose(Boolean disposing)
   at System.IO.Stream.Close()
   at AspNetCore.DataProtection.Aws.S3.S3XmlRepository.<GetElementFromKey>d__12.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at AspNetCore.DataProtection.Aws.S3.S3XmlRepository.<GetAllElementsAsync>d__11.MoveNext()<---
</code></pre>
</details>


Just to make sure, I also ran the [integration tests](https://github.com/hotchkj/AspNetCore.DataProtection.Aws/blob/master/integrate/AspNetCore.DataProtection.Aws.IntegrationTests/S3IntegrationTests.cs) pointing to my Localstack instance, and the tests without compresion passed, and those with compression failed.
<details>
<summary>Interestingly, it's the GZIP inflater that blows up in this case (click to view)</summary>
<pre><code>
System.IO.InvalidDataException : The archive entry was compressed using an unsupported compression method.
   at System.IO.Compression.Inflater.Inflate(FlushCode flushCode)
   at System.IO.Compression.Inflater.ReadInflateOutput(Byte* bufPtr, Int32 length, FlushCode flushCode, Int32& bytesRead)
   at System.IO.Compression.Inflater.InflateVerified(Byte* bufPtr, Int32 length)
   at System.IO.Compression.Inflater.Inflate(Byte[] bytes, Int32 offset, Int32 length)
   at System.IO.Compression.DeflateStream.Read(Byte[] array, Int32 offset, Int32 count)
   at System.IO.Compression.GZipStream.Read(Byte[] array, Int32 offset, Int32 count)
   at System.Xml.XmlTextReaderImpl.InitStreamInput(Uri baseUri, String baseUriStr, Stream stream, Byte[] bytes, Int32 byteCount, Encoding encoding)
   at System.Xml.XmlTextReaderImpl.FinishInitStream()
   at System.Xml.XmlTextReaderImpl..ctor(Stream stream, Byte[] bytes, Int32 byteCount, XmlReaderSettings settings, Uri baseUri, String baseUriStr, XmlParserContext context, Boolean closeInput)
   at System.Xml.XmlReaderSettings.CreateReader(Stream input, Uri baseUri, String baseUriString, XmlParserContext inputContext)
   at System.Xml.XmlReader.Create(Stream input, XmlReaderSettings settings)
   at System.Xml.Linq.XElement.Load(Stream stream, LoadOptions options)
   at System.Xml.Linq.XElement.Load(Stream stream)
   at AspNetCore.DataProtection.Aws.S3.S3XmlRepository.<GetElementFromKey>d__12.MoveNext() in C:\Repos\AspNetCore.DataProtection.Aws\src\AspNetCore.DataProtection.Aws.S3\S3XmlRepository.cs:line 189
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at AspNetCore.DataProtection.Aws.S3.S3XmlRepository.<GetAllElementsAsync>d__11.MoveNext() in C:\Repos\AspNetCore.DataProtection.Aws\src\AspNetCore.DataProtection.Aws.S3\S3XmlRepository.cs:line 144
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at AspNetCore.DataProtection.Aws.IntegrationTests.S3IntegrationTests.<ExpectCompressedStoreRetrieveToSucceed>d__12.MoveNext() in C:\Repos\AspNetCore.DataProtection.Aws\integrate\AspNetCore.DataProtection.Aws.IntegrationTests\S3IntegrationTests.cs:line 93
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
</code></pre>
</details>

It seems pretty likely to me this a Localstack issue, but I'm not really sure where to start figuring that one out. It's not a blocker for me as I can reliably get it working by disabling compression.

If you do want to investigate, I just changed the s3 client config to point at my docker instance of Localstack:

```csharp
s3Client = new AmazonS3Client(new AnonymousAWSCredentials(), new AmazonS3Config
{
    ServiceURL = "http://192.168.06:4572", // VM running docker + localstack
    ForcePathStyle = true
});
```

I ran Localstack using:

```
docker run -d -p 4572:4572 -p 8010:8010 --name localstack-s3 --restart=unless-stopped -e SERVICES=s3 -e DEFAULT_REGION=eu-west-1 -e HOSTNAME=localhost -e HOSTNAME_EXTERNAL=s3.local -e PORT_WEB_UI=8010 localstack/localstack
```

and created the test bucket using the AWS cli:

```
 aws --endpoint-url=http://localhost:4572 --region=eu-west-1 s3 mb s3://hotchkj-dataprotection-s3-integration-tests-eu-west-1
```

Thanks
